From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stransky <stransky@redhat.com>
Date: Tue, 29 Mar 2022 11:59:50 +0000
Subject: [PATCH] [sailfishos][gecko] Bug 1761471 [FFmpeg 5.0] Get frame color
 range and color space directly r=alwu

FFmpeg 5.0 is missing av_frame_get_color_range() and av_frame_get_colorspace() so we need to get that directly from AVFrame.

Differential Revision: https://phabricator.services.mozilla.com/D142064
---
 .../platforms/ffmpeg/FFmpegLibWrapper.cpp     |  8 ++-
 .../platforms/ffmpeg/FFmpegVideoDecoder.cpp   | 51 ++++++++++++-------
 .../platforms/ffmpeg/FFmpegVideoDecoder.h     |  1 +
 3 files changed, 39 insertions(+), 21 deletions(-)

diff --git a/dom/media/platforms/ffmpeg/FFmpegLibWrapper.cpp b/dom/media/platforms/ffmpeg/FFmpegLibWrapper.cpp
index ffc1c2a2e653..d698c5321444 100644
--- a/dom/media/platforms/ffmpeg/FFmpegLibWrapper.cpp
+++ b/dom/media/platforms/ffmpeg/FFmpegLibWrapper.cpp
@@ -171,8 +171,12 @@ FFmpegLibWrapper::LinkResult FFmpegLibWrapper::Link() {
   AV_FUNC(av_frame_unref,
           (AV_FUNC_AVUTIL_55 | AV_FUNC_AVUTIL_56 | AV_FUNC_AVUTIL_57 |
            AV_FUNC_AVUTIL_58 | AV_FUNC_AVUTIL_59))
-  AV_FUNC_OPTION(av_frame_get_colorspace, AV_FUNC_AVUTIL_ALL)
-  AV_FUNC_OPTION(av_frame_get_color_range, AV_FUNC_AVUTIL_ALL)
+  AV_FUNC_OPTION(av_frame_get_colorspace,
+                 AV_FUNC_AVUTIL_55 | AV_FUNC_AVUTIL_56 | AV_FUNC_AVUTIL_57 |
+                     AV_FUNC_AVUTIL_58)
+  AV_FUNC_OPTION(av_frame_get_color_range,
+                 AV_FUNC_AVUTIL_55 | AV_FUNC_AVUTIL_56 | AV_FUNC_AVUTIL_57 |
+                     AV_FUNC_AVUTIL_58)
 #ifdef MOZ_WAYLAND
   AV_FUNC_OPTION_SILENT(avcodec_get_hw_config, AV_FUNC_58 | AV_FUNC_59)
   AV_FUNC_OPTION_SILENT(av_hwdevice_ctx_init, AV_FUNC_58 | AV_FUNC_59)
diff --git a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
index 7effcc5c0a0c..fdb33247bdb7 100644
--- a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
+++ b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
@@ -593,23 +593,41 @@ MediaResult FFmpegVideoDecoder<LIBAV_VER>::DoDecode(
 }
 
 gfx::YUVColorSpace FFmpegVideoDecoder<LIBAV_VER>::GetFrameColorSpace() const {
+  AVColorSpace colorSpace = AVCOL_SPC_UNSPECIFIED;
+#if LIBAVCODEC_VERSION_MAJOR > 58
+  colorSpace = mFrame->colorspace;
+#else
   if (mLib->av_frame_get_colorspace) {
-    switch (mLib->av_frame_get_colorspace(mFrame)) {
+    colorSpace = (AVColorSpace)mLib->av_frame_get_colorspace(mFrame);
+  }
+#endif
+  switch (colorSpace) {
 #if LIBAVCODEC_VERSION_MAJOR >= 55
-      case AVCOL_SPC_BT2020_NCL:
-      case AVCOL_SPC_BT2020_CL:
-        return gfx::YUVColorSpace::BT2020;
+    case AVCOL_SPC_BT2020_NCL:
+    case AVCOL_SPC_BT2020_CL:
+      return gfx::YUVColorSpace::BT2020;
 #endif
-      case AVCOL_SPC_BT709:
-        return gfx::YUVColorSpace::BT709;
-      case AVCOL_SPC_SMPTE170M:
-      case AVCOL_SPC_BT470BG:
-        return gfx::YUVColorSpace::BT601;
-      default:
-        break;
-    }
+    case AVCOL_SPC_BT709:
+      return gfx::YUVColorSpace::BT709;
+    case AVCOL_SPC_SMPTE170M:
+    case AVCOL_SPC_BT470BG:
+      return gfx::YUVColorSpace::BT601;
+    default:
+      return DefaultColorSpace({mFrame->width, mFrame->height});
   }
-  return DefaultColorSpace({mFrame->width, mFrame->height});
+}
+
+gfx::ColorRange FFmpegVideoDecoder<LIBAV_VER>::GetFrameColorRange() const {
+  AVColorRange range = AVCOL_RANGE_UNSPECIFIED;
+#if LIBAVCODEC_VERSION_MAJOR > 58
+  range = mFrame->color_range;
+#else
+  if (mLib->av_frame_get_color_range) {
+    range = (AVColorRange)mLib->av_frame_get_color_range(mFrame);
+  }
+#endif
+  return range == AVCOL_RANGE_JPEG ? gfx::ColorRange::FULL
+                                   : gfx::ColorRange::LIMITED;
 }
 
 MediaResult FFmpegVideoDecoder<LIBAV_VER>::CreateImage(
@@ -679,12 +697,7 @@ MediaResult FFmpegVideoDecoder<LIBAV_VER>::CreateImage(
 #endif
   }
   b.mYUVColorSpace = GetFrameColorSpace();
-
-  if (mLib->av_frame_get_color_range) {
-    auto range = mLib->av_frame_get_color_range(mFrame);
-    b.mColorRange = range == AVCOL_RANGE_JPEG ? gfx::ColorRange::FULL
-                                              : gfx::ColorRange::LIMITED;
-  }
+  b.mColorRange = GetFrameColorRange();
 
   RefPtr<VideoData> v = VideoData::CreateAndCopyData(
       mInfo, mImageContainer, aOffset, TimeUnit::FromMicroseconds(aPts),
diff --git a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.h b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.h
index 7723b68e8314..f95b872ca21d 100644
--- a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.h
+++ b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.h
@@ -74,6 +74,7 @@ class FFmpegVideoDecoder<LIBAV_VER>
 #endif
   }
   gfx::YUVColorSpace GetFrameColorSpace() const;
+  gfx::ColorRange GetFrameColorRange() const;
 
   MediaResult CreateImage(int64_t aOffset, int64_t aPts, int64_t aDuration,
                           MediaDataDecoder::DecodedData& aResults) const;
